## Basic Python - Project <a id='intro'></a>

## Introduction <a id='intro'></a>
In this project, you will work with data from the entertainment industry. You will study a dataset with records on movies and shows. The research will focus on the "Golden Age" of television, which began in 1999 with the release of *The Sopranos* and is still ongoing.

The aim of this project is to investigate how the number of votes a title receives impacts its ratings. The assumption is that highly-rated shows (we will focus on TV shows, ignoring movies) released during the "Golden Age" of television also have the most votes.

### Stages 
Data on movies and shows is stored in the `/datasets/movies_and_shows.csv` file. There is no information about the quality of the data, so you will need to explore it before doing the analysis.

First, you'll evaluate the quality of the data and see whether its issues are significant. Then, during data preprocessing, you will try to account for the most critical problems.
 
Your project will consist of three stages:
 1. Data overview
 2. Data preprocessing
 3. Data analysis

## Stage 1. Data overview <a id='data_review'></a>

Open and explore the data.

You'll need `pandas`, so import it.

In [ ]:
# importing pandas
import pandas as pd

Read the `movies_and_shows.csv` file from the `datasets` folder and save it in the `df` variable:

In [ ]:
# reading the files and storing them to df
df = pd.read_csv('/datasets/movies_and_shows.csv')

Print the first 10 table rows:

In [ ]:
# obtaining the first 10 rows from the df table
# hint: you can use head() and tail() in Jupyter Notebook without wrapping them into print()
df.head(10)

name                      Character   r0le        TITLE   Type  \
0   Robert De Niro                  Travis Bickle  ACTOR  Taxi Driver  MOVIE   
1     Jodie Foster                  Iris Steensma  ACTOR  Taxi Driver  MOVIE   
2    Albert Brooks                            Tom  ACTOR  Taxi Driver  MOVIE   
3    Harvey Keitel        Matthew 'Sport' Higgins  ACTOR  Taxi Driver  MOVIE   
4  Cybill Shepherd                          Betsy  ACTOR  Taxi Driver  MOVIE   
5      Peter Boyle                         Wizard  ACTOR  Taxi Driver  MOVIE   
6   Leonard Harris      Senator Charles Palantine  ACTOR  Taxi Driver  MOVIE   
7   Diahnne Abbott                Concession Girl  ACTOR  Taxi Driver  MOVIE   
8      Gino Ardito             Policeman at Rally  ACTOR  Taxi Driver  MOVIE   
9  Martin Scorsese  Passenger Watching Silhouette  ACTOR  Taxi Driver  MOVIE   

   release Year              genres  imdb sc0re  imdb v0tes  
0          1976  ['drama', 'crime']         8.2    808582.0  
1          1976  ['drama', 'crime']         8.2    808582.0  
2          1976  ['drama', 'crime']         8.2    808582.0  
3          1976  ['drama', 'crime']         8.2    808582.0  
4          1976  ['drama', 'crime']         8.2    808582.0  
5          1976  ['drama', 'crime']         8.2    808582.0  
6          1976  ['drama', 'crime']         8.2    808582.0  
7          1976  ['drama', 'crime']         8.2    808582.0  
8          1976  ['drama', 'crime']         8.2    808582.0  
9          1976  ['drama', 'crime']         8.2    808582.0

Obtain the general information about the table with one command:

In [ ]:
# obtaining general information about the data in df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85579 entries, 0 to 85578
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0      name       85579 non-null  object 
 1   Character     85579 non-null  object 
 2   r0le          85579 non-null  object 
 3   TITLE         85578 non-null  object 
 4     Type        85579 non-null  object 
 5   release Year  85579 non-null  int64  
 6   genres        85579 non-null  object 
 7   imdb sc0re    80970 non-null  float64
 8   imdb v0tes    80853 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 5.9+ MB


The table contains nine columns. The majority store the same data type: object. The only exceptions are `'release Year'` (int64 type), `'imdb sc0re'` (float64 type) and `'imdb v0tes'` (float64 type). Scores and votes will be used in our analysis, so it's important to verify that they are present in the dataframe in the appropriate numeric format. Three columns (`'TITLE'`, `'imdb sc0re'` and `'imdb v0tes'`) have missing values.

According to the documentation:
- `'name'` — actor/director's name and last name
- `'Character'` — character played (for actors)
- `'r0le '` — the person's contribution to the title (it can be in the capacity of either actor or director)
- `'TITLE '` — title of the movie (show)
- `'  Type'` — show or movie
- `'release Year'` — year when movie (show) was released
- `'genres'` — list of genres under which the movie (show) falls
- `'imdb sc0re'` — score on IMDb
- `'imdb v0tes'` — votes on IMDb

We can see three issues with the column names:
1. Some names are uppercase, while others are lowercase.
2. There are names containing whitespace.
3. A few column names have digit '0' instead of letter 'o'. 


### Conclusions <a id='data_review_conclusions'></a> 

Each row in the table stores data about a movie or show. The columns can be divided into two categories: the first is about the roles held by different people who worked on the movie or show (role, name of the actor or director, and character if the row is about an actor); the second category is information about the movie or show itself (title, release year, genre, imdb figures).

It's clear that there is sufficient data to do the analysis and evaluate our assumption. However, to move forward, we need to preprocess the data.

## Stage 2. Data preprocessing <a id='data_preprocessing'></a>
Correct the formatting in the column headers and deal with the missing values. Then, check whether there are duplicates in the data.

In [ ]:
# the list of column names in the df table
df.columns

Index(['   name', 'Character', 'r0le', 'TITLE', '  Type', 'release Year',
       'genres', 'imdb sc0re', 'imdb v0tes'],
      dtype='object')

Change the column names according to the rules of good style:
* If the name has several words, use snake_case
* All characters must be lowercase
* Remove whitespace
* Replace zero with letter 'o'

In [ ]:
# renaming columns
df = df.rename(columns={
    '   name' : 'name',
    'Character': 'character',
    'r0le' : 'role',
    'TITLE': 'title',
    '  Type': 'type',
    'release Year': 'release_year',
    'imdb sc0re': 'imdb_score',
    'imdb v0tes': 'imdb_votes',
})

Check the result. Print the names of the columns once more:

In [ ]:
# checking result: the list of column names
df.columns

Index(['name', 'character', 'role', 'title', 'type', 'release_year', 'genres',
       'imdb_score', 'imdb_votes'],
      dtype='object')

### Missing values <a id='missing_values'></a>
First, find the number of missing values in the table. To do so, combine two `pandas` methods:

In [ ]:
# calculating missing values
df.isna().sum()

name               0
character          0
role               0
title              1
type               0
release_year       0
genres             0
imdb_score      4609
imdb_votes      4726
dtype: int64

We identified missing values in several columns. While the missing value in `'title'` isn't critical, missing values in `'imdb_score'` and `'imdb_votes'` affect around 6% of the data, which could impact our analysis. To ensure data integrity, we'll drop all rows with missing values.

In [ ]:
# dropping rows where columns with scores, and votes have missing values
df = df.dropna(axis='rows')

Make sure the table doesn't contain any more missing values. Count the missing values again.

In [ ]:
# counting missing values
df.isna().sum()

name            0
character       0
role            0
title           0
type            0
release_year    0
genres          0
imdb_score      0
imdb_votes      0
dtype: int64

### Duplicates <a id='duplicates'></a>
Find the number of duplicate rows in the table using one command:

In [ ]:
#counting duplicate rows
duplicates = df.duplicated().sum()
duplicates

6994

Review the duplicate rows to determine if removing them would distort our dataset.

In [ ]:
# Produce table with duplicates (with original rows included) and review last 5 rows
duplicated_rows = df[df.duplicated(keep=False)]
duplicated_rows.head()
#should this be .tail() for last 5 

name                  character      role  \
7560      Philip Greene  Baseball Fan (uncredited)     ACTOR   
7561      Philip Greene  Baseball Fan (uncredited)     ACTOR   
14502          Dan Levy                   Reporter     ACTOR   
14512          Dan Levy                   Reporter     ACTOR   
18951  Nicolas Le Nev??                    unknown  DIRECTOR   

                                 title   type  release_year  \
7560                   How Do You Know  MOVIE          2010   
7561                   How Do You Know  MOVIE          2010   
14502  A Very Harold & Kumar Christmas  MOVIE          2011   
14512  A Very Harold & Kumar Christmas  MOVIE          2011   
18951                       Sammy & Co   SHOW          2014   

                                 genres  imdb_score  imdb_votes  
7560     ['comedy', 'drama', 'romance']         5.4     50383.0  
7561     ['comedy', 'drama', 'romance']         5.4     50383.0  
14502  ['comedy', 'fantasy', 'romance']         6.2     69562.0  
14512  ['comedy', 'fantasy', 'romance']         6.2     69562.0  
18951         ['animation', 'european']         5.7        31.0

There are two clear duplicates in the printed rows. We can safely remove them.
Call the `pandas` method for getting rid of duplicate rows:

In [ ]:
# removing duplicate rows
df = df.drop_duplicates()

Check for duplicate rows once more to make sure you have removed all of them:

In [ ]:
# checking for duplicates
duplicates = df.duplicated().sum()
duplicates

0

Now get rid of implicit duplicates in the `'type'` column. For example, the string `'SHOW'` can be written in different ways. These kinds of errors will also affect the result.

Print a list of unique `'type'` names, sorted in alphabetical order. To do so:
* Retrieve the intended dataframe column 
* Apply a sorting method to it
* For the sorted column, call the method that will return all unique column values

In [ ]:
# viewing unique type names
unique_type = df['type'].unique()
unique_type

array(['MOVIE', 'the movie', 'tv show', 'shows', 'movies', 'tv shows',
       'tv series', 'tv', 'SHOW'], dtype=object)

Look through the list to find implicit duplicates of `'show'` (`'movie'` duplicates will be ignored since the assumption is about shows). These could be names written incorrectly or alternative names of the same genre.

You will see the following implicit duplicates:
* `'shows'`
* `'SHOW'`
* `'tv show'`
* `'tv shows'`
* `'tv series'`
* `'tv'`

To get rid of them, declare the function `replace_wrong_show()` with two parameters: 
* `wrong_shows_list=` — the list of duplicates
* `correct_show=` — the string with the correct value

The function should correct the names in the `'type'` column from the `df` table (i.e., replace each value from the `wrong_shows_list` list with the value in `correct_show`).

In [ ]:
# function for replacing implicit duplicates
wrong_shows_list = ['tv show', 'shows', 'tv shows', 'tv series', 'tv', 'SHOW']
correct_show = 'show'
wrong_movie_list = 'MOVIE', 'the movie', 'movies'
correct_movie = 'movie'

def replace_wrong_show(wrong_shows_list, correct_show):
    for name in wrong_shows_list:
        df['type']= df['type'].replace(name, correct_show)
        
def replace_wrong_movie(wrong_movie_list, correct_movie):
    for name in wrong_movie_list:
        df['type']= df['type'].replace(name, correct_movie)

Call `replace_wrong_show()` and pass it arguments so that it clears implicit duplicates and replaces them with `SHOW`:

In [ ]:
# removing implicit duplicates
replace_wrong_show(wrong_shows_list, correct_show)
replace_wrong_movie(wrong_movie_list, correct_movie)

Make sure the duplicate names are removed. Print the list of unique values from the `'type'` column:

In [ ]:
# viewing unique genre names
df['type'].sort_values().unique()

array(['movie', 'show'], dtype=object)

### Conclusions <a id='data_preprocessing_conclusions'></a>
We detected three issues with the data:

- Incorrect header styles
- Missing values
- Duplicate rows and implicit duplicates

The headers have been cleaned up to make processing the table simpler.

All rows with missing values have been removed. 

The absence of duplicates will make the results more precise and easier to understand.

Now we can move on to our analysis of the prepared data.

## Stage 3. Data analysis <a id='hypotheses'></a>

Based on the previous project stages, you can now define how the assumption will be checked. Calculate the average amount of votes for each score (this data is available in the `imdb_score` and `imdb_votes` columns), and then check how these averages relate to each other. If the averages for shows with the highest scores are bigger than those for shows with lower scores, the assumption appears to be true.

Based on this, complete the following steps:

- Filter the dataframe to only include shows released in 1999 or later.
- Group scores into buckets by rounding the values of the appropriate column (a set of 1-10 integers will help us make the outcome of our calculations more evident without damaging the quality of our research).
- Identify outliers among scores based on their number of votes, and exclude scores with few votes.
- Calculate the average votes for each score and check whether the assumption matches the results.

To filter the dataframe and only include shows released in 1999 or later, you will take two steps. First, keep only titles published in 1999 or later in our dataframe. Then, filter the table to only contain shows (movies will be removed).

In [ ]:
# using conditional indexing modify df so it has only titles released after 1999 (with 1999 included)
# give the slice of dataframe new name
titles_after_1999 = df[df['release_year'] >=1999]

In [ ]:
# repeat conditional indexing so df has only shows (movies are removed as result)
shows_after_1999 = after_1999_titles[after_1999_titles['type'] == 'show']

The scores that are to be grouped should be rounded. For instance, titles with scores like 7.8, 8.1, and 8.3 will all be placed in the same bucket with a score of 8.

In [ ]:
# rounding column with scores
shows_after_1999['imdb_score'] = df['imdb_score'].round(0)
#shows_after_1999['imdb_score'] = shows_after_1999['imdb_score'].round(0).copy()
    #checking the outcome with tail()
shows_after_1999['imdb_score'].tail()

/tmp/ipykernel_32/1914999057.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  shows_after_1999['imdb_score'] = df['imdb_score'].round(0)


85433    8.0
85434    8.0
85435    8.0
85436    8.0
85437    8.0
Name: imdb_score, dtype: float64

It is now time to identify outliers based on the number of votes.

In [ ]:
# Use groupby() for scores and count all unique values in each group, print the result
shows_after_1999.groupby('imdb_score').nunique()


name  character  role  title  type  release_year  genres  \
imdb_score                                                             
2.0           24         18     2      5     1             4       5   
3.0           27         27     2      4     1             3       4   
4.0          174        108     2     30     1            13      19   
5.0          581        419     2     94     1            12      68   
6.0         2365       1896     2    347     1            19     212   
7.0         4342       3813     2    593     1            24     304   
8.0         4194       4028     2    539     1            23     329   
9.0          542        498     2     65     1            20      50   
10.0           8          8     1      1     1             1       1   

            imdb_votes  
imdb_score              
2.0                  5  
3.0                  4  
4.0                 30  
5.0                 91  
6.0                329  
7.0                568  
8.0                527  
9.0                 65  
10.0                 1

Based on the aggregation performed, it is evident that scores 2 (24 voted shows), 3 (27 voted shows), and 10 (only 8 voted shows) are outliers. There isn't enough data for these scores for the average number of votes to be meaningful.

To obtain the mean numbers of votes for the selected scores (we identified a range of 4-9 as acceptable), use conditional filtering and grouping.

In [ ]:
# filter dataframe using two conditions (scores to be in the range 4-9)
shows_after_1999 = shows_after_1999[shows_after_1999['imdb_score'] >=4.0]
shows_after_1999 = shows_after_1999[shows_after_1999['imdb_score'] <=9.0]
# group scores and corresponding average number of votes, reset index and print the result
grouped_df = shows_after_1999.groupby('imdb_score')['imdb_votes'].mean().reset_index()
grouped_df

imdb_score     imdb_votes
0         4.0    5277.583333
1         5.0    3143.942568
2         6.0    3481.717322
3         7.0    8727.068211
4         8.0   30299.460967
5         9.0  126904.109515

Now for the final step! Round the column with the averages, rename both columns, and print the dataframe in descending order.

In [ ]:
# round column with averages
grouped_df['imdb_votes'] = grouped_df['imdb_votes'].round(0)
# rename columns
grouped_df = grouped_df.rename(columns={'imdb_score': 'Score', 'imdb_votes': 'Average Votes'})
# print dataframe in descending order
grouped_df = grouped_df.sort_values(by='Average Votes', ascending=False)

The assumption macthes the analysis: the shows with the top 3 scores have the most amounts of votes.

In [ ]:
grouped_df.head(6)


Score  Average Votes
5    9.0       126904.0
4    8.0        30299.0
3    7.0         8727.0
0    4.0         5278.0
2    6.0         3482.0
1    5.0         3144.0

In [ ]:
grouped_df = grouped_df.sort_values(by='Average Votes', ascending=False)
grouped_df.tail(6)

Score  Average Votes
5    9.0       126904.0
4    8.0        30299.0
3    7.0         8727.0
0    4.0         5278.0
2    6.0         3482.0
1    5.0         3144.0

In [ ]:
grouped_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6 entries, 5 to 1
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Score          6 non-null      float64
 1   Average Votes  6 non-null      float64
dtypes: float64(2)
memory usage: 144.0 bytes


## Conclusion <a id='hypotheses'></a>

The research done confirms that highly-rated shows released during the "Golden Age" of television also have the most votes. While shows with score 4 have more votes than ones with scores 5 and 6, the top three (scores 7-9) have the largest number. The data studied represents around 94% of the original set, so we can be confident in our findings.